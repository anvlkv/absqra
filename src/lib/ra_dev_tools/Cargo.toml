[package]
name = "ra_dev_tools"
version = "0.1.0"
authors = ["a.nvlkv <a.nvlkv@protonmail.ch>"]
edition = "2018"
build = "build.rs"

[build-dependencies]
build-deps = "0.1.4"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = "0.16.0"
regex = "1.3.9"
example_files_macro = {path="./src/lib/example_files_macro"}
files_iterator = {path="./src/lib/files_iterator"}
