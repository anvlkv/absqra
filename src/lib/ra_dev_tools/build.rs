extern crate build_deps;

fn main() {
    // Enumerate files in sub-folder "examples/*", being relevant for the test-generation (as example)
    // If function returns with error, exit with error message.
    build_deps::rerun_if_changed_paths( "../../../examples/*/*" ).unwrap();

    // Adding the parent directory "examples" to the watch-list will capture new-files being added
    build_deps::rerun_if_changed_paths( "../../../examples/*" ).unwrap();
}