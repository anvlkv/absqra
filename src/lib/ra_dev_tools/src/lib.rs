pub extern crate insta;
pub use files_iterator::for_each_ra_example_file;

#[allow(unused_attributes)]
#[macro_use] pub use example_files_macro::make_example_tests;

#[cfg(test)]
mod test {
    use super::make_example_tests;

    #[make_example_tests]
    #[test]
    fn it_works_with(content: String, file_name: String) {
        println!("didn't panic with content: \n[{}] of file_name: {}", content, file_name);
    }
}