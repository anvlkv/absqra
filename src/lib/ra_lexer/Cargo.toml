[package]
name = "ra_lexer"
version = "0.1.0"
authors = ["a.nvlkv <a.nvlkv@protonmail.ch>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
failure = "0.1.8"
failure_derive = "0.1.8"
serde = "1.0.114"
serde_json = "1.0.56"
serde_derive = "1.0.114"

[dev-dependencies]
ra_dev_tools = {path = "../ra_dev_tools"}
