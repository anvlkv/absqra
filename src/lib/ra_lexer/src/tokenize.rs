use super::*;

pub fn tokenize<'a>(input: &'a str) -> impl Iterator<Item = Result<RaToken, LexerError>> + 'a {
    Cursor::from(input)
    // tokenize_cursor(Cursor::new(input, Position(1, 0), 0, 0)).map(|lexer_result| lexer_result?.try_into())
}