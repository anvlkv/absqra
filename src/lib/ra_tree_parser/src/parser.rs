use super::*;
use ra_lexer::errors::LexerError;

pub fn parse<'a>(
    mut tokens_stream: impl Iterator<Item = Result<RaToken, LexerError>>
) -> Result<RaTree, (Vec<LexerError>, Option<RaTree>)> {

    let mut tree = RaTree::new();

    let mut errors: Vec<LexerError> = Vec::new();

    while let Some(token_result) = tokens_stream.next() {
        match token_result {
            Ok(token) => {
                tree.add_token(token);
            },
            Err(e) => {
                errors.push(e);
            }
        }
    }
    if errors.len() > 0 {
        Err((errors, Some(tree)))
    }
    else {
        Ok(tree)
    }
}
