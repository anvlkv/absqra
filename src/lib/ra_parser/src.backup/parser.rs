use crate::expressions::context_expression::{ContextExpression, ContextExpressionMemberKind, ContextExpressionMember};
use ra_lexer::token::{RaToken, TokenKind};
use ra_lexer::tokenize;
use ra_lexer::errors::LexerError;

use super::block::{Block, BlockKind};
use super::cursor::Cursor;
use super::errors::ParserError;
use super::expressions::traits::{Leveled, Expandable, Positioned};
use failure::Backtrace;

pub fn parse<'a>(input: &'a str) -> Result<Block<'a>, (Vec<ParserError>, Block<'a>)> {
    let stream = tokenize(input);
    
    let mut errors: Vec<ParserError> = Vec::new();

    let program = {
        let mut block = Block::new(RaToken::default()).unwrap();

        let mut tokens_cursor = Cursor::new(
            stream
            .take_while(|r| r.is_ok())
            .map(|r| r.unwrap())
            .filter(|tok| tok.kind.unwrap() != TokenKind::Comment)
        );
        loop {
            if !tokens_cursor.is_eof() {
                match tokens_cursor.advance_block() {
                    Ok(blk) => block.children.push(blk),
                    Err(es) => errors.extend(es.into_iter()),
                }
            } else {
                break;
            }
        }

        block
    };

    if errors.len() > 0 {
        Err((errors, program))
    } else {
        Ok(program)
    }
}

impl<'token, I> Cursor<'token, I>
where
    I: Iterator<Item = RaToken<'token>>,
{
    fn advance_block<'a>(&'a mut self) -> Result<Block<'token>, Vec<ParserError>> {
        let first_token = self.bump();
        let mut errors: Vec<ParserError> = Vec::new();
        match first_token {
            Some(token) => {
                self.parse_block(token)
            },
            None => {
                errors.push(ParserError::UnexpectedEndOfInput(self.position, Backtrace::new()));
                Err(errors)
            }
        }
    }

    fn parse_block(
        &mut self,
        first_token: RaToken<'token>,
    ) -> Result<Block<'token>, Vec<ParserError>> {
        let mut errors: Vec<ParserError> = Vec::new();
        let block = Block::new(first_token)?;

        if let (Some(blk), mut errs) =
            self.check_parse_block_expression(block, &mut errors)
        {
            if let (Some(blk), mut errs) = self.check_parse_block_children(blk, &mut errs) {
                if let (Some(blk), errs) = self.check_parse_union(blk, &mut errs) {
                    if errs.len() == 0 {
                        return Ok(blk);
                    }
                }
            }
        }

        Err(errors)
    }

    fn check_parse_block_expression<'errors>(
        &mut self,
        mut block: Block<'token>,
        errors: &'errors mut Vec<ParserError>,
    ) -> (Option<Block<'token>>, &'errors mut Vec<ParserError>) {
        let mut lvl = block.get_level();
        let mut line_number = (block.get_position().0).0;
        while self.first_ahead().is_some()
            && self.first_ahead().unwrap().level == lvl
            && (self.first_ahead().unwrap().position.0).0 == line_number {
                match block.clone().append_item(self.bump().unwrap()) {
                    Ok(blk) => {
                        block = blk;

                        // // TODO: ugly, move this logic elsewhere

                        if let BlockKind::ContextModification(Some(ContextExpression(target, source))) = block.clone().kind { 
                            if ContextExpressionMember::Target(ContextExpressionMemberKind::MSpecifier) == target && source.is_none() {
                                if let (Some(mut blk), mut errs) = self.check_parse_block_children(block.clone(), errors) {
                                    let close_curly_brace = self.bump().unwrap();
                                    assert_eq!(close_curly_brace.kind.unwrap(), TokenKind::CloseCurlyBrace);
                                    block = blk;
                                    lvl = close_curly_brace.level;
                                    line_number = (close_curly_brace.position.0).0;
                            //         // let updated_expression = ContextExpression(target, None).append_item(self.bump().unwrap())?;

                            //         // blk.kind = BlockKind::ContextModification(Some(updated_expression));
                            //         // self.check_parse_block_expression(blk, errs)
                                }
                            }
                            else if source.is_some() && ContextExpressionMember::Source(ContextExpressionMemberKind::MSpecifier) == source.unwrap() {
                                if let (Some(blk), mut errs) = self.check_parse_block_children(block.clone(), errors) {
                                    assert_eq!(self.bump().unwrap().kind.unwrap(), TokenKind::CloseCurlyBrace);
                                    assert_eq!(self.bump().unwrap().kind.unwrap(), TokenKind::CloseCurlyBrace);
                                    block = blk;
                                    // return Ok(blk);
                                }
                            }
                            println!("{:?}", block);
                        }

                        // // end TODO
                    },
                    Err(e) => {
                        errors.push(e)
                    }
                }
        }

        (Some(block), errors)
    }

    fn check_parse_block_children<'errors>(
        &mut self,
        mut block: Block<'token>,
        errors: &'errors mut Vec<ParserError>,
    ) -> (Option<Block<'token>>, &'errors mut Vec<ParserError>) {
        
        while !self.is_eof() && self.first_ahead().unwrap().level == block.get_level() + 1 {
            match self.advance_block() {
                Ok(child) => {
                    block.children.push(child);
                }
                Err(errs) => {
                    errors.extend(errs);
                }
            }
        }

        if errors.len() == 0 {
            (Some(block), errors)
        } else {
            (None, errors)
        }
    }

    fn check_parse_union<'errors>(
        &mut self,
        block: Block<'token>,
        errors: &'errors mut Vec<ParserError>,
    ) -> (Option<Block<'token>>, &'errors mut Vec<ParserError>) {
        let mut union_block = block.clone();
        let lvl = block.get_level();
        let mut union_size = 0;
        while !self.is_eof()
            && self.first_ahead().unwrap().kind.unwrap() == TokenKind::Coma
            && self.first_ahead().unwrap().level == lvl
        {
            assert!(self.bump().unwrap().kind.unwrap() == TokenKind::Coma);
            match self.advance_block() {
                Ok(blk) => {
                    union_size += 1;
                    union_block.children.push(blk);
                }
                Err(errs) => {
                    errors.extend(errs);
                }
            }
        }
        
        if union_size > 0 {
            union_block.kind = BlockKind::Union(union_size);
            (Some(union_block), errors)
        } else {
            (Some(block), errors)
        }
    }
}
