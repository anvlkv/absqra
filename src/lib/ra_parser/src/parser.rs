use crate::errors::ParserError;
use crate::block::Block;
use crate::parsed_by_token::ParsedByToken;
use ra_lexer::tokenize;
use ra_lexer::token::TokenKind;

pub fn parse<'a>(input: &'a str) -> Result<Block, (Vec<ParserError>, Option<Block>)> {
    let mut errors: Vec<ParserError> = Vec::new();
    let mut tokens_stream = tokenize(input);
    
    let mut block = {
        match Block::new_program() {
            Ok(b) => b,
            Err(e) => {
                errors.extend(e);
                return Err((errors, None));
            }
        }
    };

    while let Some(token_result) = tokens_stream.next() {
        match token_result {
            Ok(token) => {
                if token.kind == TokenKind::Comment {
                    continue;
                }
                else {
                    match block.clone().append_token(token) {
                        Ok(blk) => block=blk,
                        Err(e) => errors.extend(e)
                    }
                }
            },
            Err(e) => {
                errors.push(e.into());
            }
        }
    }

    if errors.len() == 0 {
        Ok(*block)
    }
    else {
        Err((errors, Some(*block)))
    }
}