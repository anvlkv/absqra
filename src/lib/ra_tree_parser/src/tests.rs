
mod lib {
    use ra_dev_tools::insta::{assert_snapshot, assert_debug_snapshot};    
    use ra_dev_tools::{make_example_tests};
    use ra_lexer::tokenize;
    use crate::parser::parse;

    #[make_example_tests]
    #[test]
    fn it_should_match_snapshots(contents: String, file_name: String) {
        let err_messages_file_name = format!("{}__{}",file_name, "ERR__MESSAGES");
        let errors_file_name = format!("{}__{}", file_name, "ERR");
        let tree = {
            let mut tokens_iter = tokenize(&contents);
            match parse(tokens_iter) {
                Ok(tree) => {
                    tree
                },
                Err((errors, parsed)) => {
                    let formatted_errors: String = {
                        errors.into_iter().map(|e| format!("{}", e)).collect::<Vec<String>>().join("\n")
                    };
                    
                    assert_snapshot!(err_messages_file_name, formatted_errors);
                    assert_debug_snapshot!(errors_file_name, parsed);


                    panic!("failed to parse example {:?}", file_name);
                }
            }
        };
        assert_debug_snapshot!(file_name, tree)       
    }
}
