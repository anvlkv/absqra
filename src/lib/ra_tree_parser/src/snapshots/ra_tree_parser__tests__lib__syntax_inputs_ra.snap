---
source: src/lib/ra_tree_parser/src/tests.rs
expression: tree
---
[
  [
    "Root",
    [
      [
        "Block",
        [
          [
            {
              "Token": {
                "kind": "Greater",
                "position": [
                  [
                    1,
                    1
                  ],
                  [
                    1,
                    2
                  ]
                ],
                "level": 1
              }
            }
          ],
          [
            {
              "Token": {
                "kind": {
                  "Comment": [
                    " simply a text input",
                    false
                  ]
                },
                "position": [
                  [
                    1,
                    3
                  ],
                  [
                    1,
                    25
                  ]
                ],
                "level": 1
              }
            }
          ]
        ]
      ],
      [
        "Block",
        [
          [
            {
              "Token": {
                "kind": "Plus",
                "position": [
                  [
                    2,
                    1
                  ],
                  [
                    2,
                    2
                  ]
                ],
                "level": 1
              }
            }
          ],
          [
            {
              "Token": {
                "kind": "Greater",
                "position": [
                  [
                    2,
                    2
                  ],
                  [
                    2,
                    3
                  ]
                ],
                "level": 1
              }
            }
          ],
          [
            {
              "Token": {
                "kind": {
                  "Comment": [
                    " simply an array input",
                    false
                  ]
                },
                "position": [
                  [
                    2,
                    4
                  ],
                  [
                    2,
                    28
                  ]
                ],
                "level": 1
              }
            }
          ]
        ]
      ],
      [
        "Block",
        [
          [
            {
              "Token": {
                "kind": "Greater",
                "position": [
                  [
                    3,
                    1
                  ],
                  [
                    3,
                    2
                  ]
                ],
                "level": 1
              }
            }
          ],
          [
            "Block",
            [
              [
                {
                  "Token": {
                    "kind": "Exclamation",
                    "position": [
                      [
                        4,
                        5
                      ],
                      [
                        4,
                        6
                      ]
                    ],
                    "level": 2
                  }
                }
              ],
              [
                {
                  "Token": {
                    "kind": {
                      "Identifier": "type"
                    },
                    "position": [
                      [
                        4,
                        6
                      ],
                      [
                        4,
                        10
                      ]
                    ],
                    "level": 2
                  }
                }
              ],
              [
                {
                  "Token": {
                    "kind": "Colon",
                    "position": [
                      [
                        4,
                        10
                      ],
                      [
                        4,
                        11
                      ]
                    ],
                    "level": 2
                  }
                }
              ],
              [
                {
                  "Token": {
                    "kind": {
                      "Identifier": "number"
                    },
                    "position": [
                      [
                        4,
                        11
                      ],
                      [
                        4,
                        17
                      ]
                    ],
                    "level": 2
                  }
                }
              ],
              [
                {
                  "Token": {
                    "kind": {
                      "Comment": [
                        " typed input",
                        false
                      ]
                    },
                    "position": [
                      [
                        4,
                        18
                      ],
                      [
                        4,
                        32
                      ]
                    ],
                    "level": 2
                  }
                }
              ]
            ]
          ]
        ]
      ],
      [
        "Block",
        [
          [
            {
              "Token": {
                "kind": "Exclamation",
                "position": [
                  [
                    5,
                    5
                  ],
                  [
                    5,
                    6
                  ]
                ],
                "level": 2
              }
            }
          ],
          [
            {
              "Token": {
                "kind": {
                  "Identifier": "max"
                },
                "position": [
                  [
                    5,
                    6
                  ],
                  [
                    5,
                    9
                  ]
                ],
                "level": 2
              }
            }
          ],
          [
            {
              "Token": {
                "kind": "Colon",
                "position": [
                  [
                    5,
                    9
                  ],
                  [
                    5,
                    10
                  ]
                ],
                "level": 2
              }
            }
          ],
          [
            {
              "Token": {
                "kind": {
                  "IntegerLiteral": 20
                },
                "position": [
                  [
                    5,
                    10
                  ],
                  [
                    5,
                    12
                  ]
                ],
                "level": 2
              }
            }
          ],
          [
            {
              "Token": {
                "kind": {
                  "Comment": [
                    " max value = 20",
                    false
                  ]
                },
                "position": [
                  [
                    5,
                    14
                  ],
                  [
                    5,
                    31
                  ]
                ],
                "level": 2
              }
            }
          ]
        ]
      ],
      [
        "Block",
        [
          [
            {
              "Token": {
                "kind": "Plus",
                "position": [
                  [
                    7,
                    1
                  ],
                  [
                    7,
                    2
                  ]
                ],
                "level": 1
              }
            }
          ],
          [
            {
              "Token": {
                "kind": "Greater",
                "position": [
                  [
                    7,
                    2
                  ],
                  [
                    7,
                    3
                  ]
                ],
                "level": 1
              }
            }
          ],
          [
            "Block",
            [
              [
                {
                  "Token": {
                    "kind": "Exclamation",
                    "position": [
                      [
                        8,
                        5
                      ],
                      [
                        8,
                        6
                      ]
                    ],
                    "level": 2
                  }
                }
              ],
              [
                {
                  "Token": {
                    "kind": {
                      "Identifier": "type"
                    },
                    "position": [
                      [
                        8,
                        6
                      ],
                      [
                        8,
                        10
                      ]
                    ],
                    "level": 2
                  }
                }
              ],
              [
                {
                  "Token": {
                    "kind": "Colon",
                    "position": [
                      [
                        8,
                        10
                      ],
                      [
                        8,
                        11
                      ]
                    ],
                    "level": 2
                  }
                }
              ],
              [
                {
                  "Token": {
                    "kind": {
                      "Identifier": "string"
                    },
                    "position": [
                      [
                        8,
                        11
                      ],
                      [
                        8,
                        17
                      ]
                    ],
                    "level": 2
                  }
                }
              ]
            ]
          ]
        ]
      ],
      [
        "Block",
        [
          [
            {
              "Token": {
                "kind": "Exclamation",
                "position": [
                  [
                    9,
                    5
                  ],
                  [
                    9,
                    6
                  ]
                ],
                "level": 2
              }
            }
          ],
          [
            {
              "Token": {
                "kind": {
                  "Identifier": "max_length"
                },
                "position": [
                  [
                    9,
                    6
                  ],
                  [
                    9,
                    16
                  ]
                ],
                "level": 2
              }
            }
          ],
          [
            {
              "Token": {
                "kind": "Colon",
                "position": [
                  [
                    9,
                    16
                  ],
                  [
                    9,
                    17
                  ]
                ],
                "level": 2
              }
            }
          ],
          [
            {
              "Token": {
                "kind": {
                  "IntegerLiteral": 3
                },
                "position": [
                  [
                    9,
                    17
                  ],
                  [
                    9,
                    18
                  ]
                ],
                "level": 2
              }
            }
          ]
        ]
      ],
      [
        "Block",
        [
          [
            {
              "Token": {
                "kind": {
                  "Comment": [
                    "\narray input of strings\neach string has max length of 3\n",
                    false
                  ]
                },
                "position": [
                  [
                    10,
                    5
                  ],
                  [
                    13,
                    7
                  ]
                ],
                "level": 2
              }
            }
          ]
        ]
      ]
    ]
  ]
]
